doc: add board terminology section to board porting guide

tejlmand opened the pull request:
Add a terminology section to the board porting guide. This provides a single location where terminologies used for boards can be found and linked to.

carlescufi commented on doc/hardware/porting/board_porting.rst:
```diff
@@ -956,3 +956,23 @@ extension.
+Terminology
+***********
+
+- board: a physical piece of hardware or an emulated board.
```
Please add an example for each of the terms

nashif commented on doc/hardware/porting/board_porting.rst:
```diff
@@ -956,3 +956,23 @@ extension.
+
+.. _board_terminology:
+
+Terminology
```
would be nicer if you used some formatting for this type of thing, so it is easy to read and find:

https://sublime-and-sphinx-guide.readthedocs.io/en/latest/glossary.html#create-a-glossary

nashif replied:
would be nice as a followup to have few diagrams showing all of this and the layering and how things relate and what type of "setups" are possible. a picture is worth 1000 words....

tejlmand replied:
> would be nicer if you used some formatting for this type of thing, so it is easy to read and find:
> 
> https://sublime-and-sphinx-guide.readthedocs.io/en/latest/glossary.html#create-a-glossary

Thanks for the link, will take a look at updating.
I just followed what we do here: https://docs.zephyrproject.org/latest/project/index.html#terminology

Perhaps long term we should try to have a uniform terminology style, fyi @kartben

nashif replied:
now that you mention this, you should remove the "/" from identifier :)

nashif replied:
and yes, we can start with a link to this table.

tejlmand replied:
> > would be nicer if you used some formatting for this type of thing, so it is easy to read and find:
> > https://sublime-and-sphinx-guide.readthedocs.io/en/latest/glossary.html#create-a-glossary
> 
> Thanks for the link, will take a look at updating.

Done

tejlmand replied:
> and yes, we can start with a link to this table.

added a reference to the section as I couldn't get referencing the table directly to work.

tejlmand replied:
> now that you mention this, you should remove the "/" from identifier :)

done

carlescufi reviewed (dismissed):
LGTM, thank you!

ajarmouni-st commented on doc/hardware/porting/board_porting.rst:
```diff
@@ -956,3 +956,77 @@ extension.
+    board identifier
+      a name which identifies the exact SoC, CPU cluster, or board
+      variant for the current build. The board identifier is separated from the
+      board name by a ``/``. Similar the SoC is also separated from the CPU
```
```suggestion
      board name by a ``/``. Similarly, the SoC is also separated from the CPU
```

ajarmouni-st commented on doc/hardware/porting/board_porting.rst:
```diff
@@ -956,3 +956,77 @@ extension.
+- ``nrf5340/cpuapp``: is the board identifier, and can be further divided into
+
+  - ``nrf5340``: The SoC, which is ``nrf5340``
+  - ``cpuapp``: The CPU cluster ``cpuapp`` which for this SoC is a single CPU
+    core of Cortex-m33. The number of cores in a CPU cluster cannot be
```
```suggestion
  - ``cpuapp``: The CPU cluster ``cpuapp``, which is a single Cortex-M33 CPU
    core for this SoC. The number of cores in a CPU cluster cannot be
```

ajarmouni-st commented on doc/hardware/porting/board_porting.rst:
```diff
@@ -956,3 +956,77 @@ extension.
+- ``ns``: a board variant,, in this case ``ns`` is a common variant name is
+  Zephyr denoting a non-secure build for boards supporting :ref:`tfm`.
+
+Not all SoCs defines CPU clusters or board variants. For example a simple board
+like the :ref:`thingy52_nrf52832` contains a single SoC with no CPU cluster and
```
```suggestion
Not all SoCs define CPU clusters or board variants. For example a simple board
like the :ref:`thingy52_nrf52832` contains a single SoC with no CPU clusters and
```

ajarmouni-st commented on doc/hardware/porting/board_porting.rst:
```diff
@@ -956,3 +956,77 @@ extension.
+read as:
+
+- ``thingy52``: This part is before the first ``/`` and is the board name.
+- ``nrf52832``: is the board identifier, and as there is no more levels then
+  it's in this case identical the SoC, which is ``nrf52832``.
```
```suggestion
- ``nrf52832``: is the board identifier, and as there are no more levels then
  it's in this case identical to the SoC, which is ``nrf52832``.
```

kartben commented on doc/hardware/porting/board_porting.rst:
```diff
@@ -956,3 +956,77 @@ extension.
+
+.. _board_terminology:
+
+Terminology
```
sorry if I missed something but this list is almost exactly a copy-paste of the "Boards, SoCs, etc." section, so I am not sure there is a reason for the two to still co-exist?

tejlmand replied:
there was a discussion on having a terminology section, see https://discord.com/channels/720317445772017664/1194671118422913114/1214257604205744158 and before comments. A proposal to which I agree, especially as it also makes it easier to reference.

But you're right that `Boards, SoCs, etc.` contains part of same, but the reason I didn't remove it from there is because I still think it makes sense to have a brief description at that location where we introduces the concepts.

At the same time, a reader will expect that a terminology section describes all important terms, as a form of quick reference. So hard to argue those term should not be there.

carlescufi replied:
I think we can live with this duplication. Longer term what we need is a linkable global Glossary instead, like we do in NCS:
https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/glossary.html

kartben replied:
> I think we can live with this duplication. Longer term what we need is a linkable global Glossary instead, like we do in NCS:
> https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/glossary.html
> 

Exactly :) What's wrong with using the glossary we already have? The text that's used in "Boards, SoCs, ..." can stay there as a legend for the diagram, but should also link to proper entries in the glossary (and this will be a good opportunity to update existing entries such as "SoC")
https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/glossary.rst

tejlmand commented on doc/hardware/porting/board_porting.rst:
```diff
@@ -956,3 +956,77 @@ extension.
+      and where a given revision needs adjustments to devicetree or Kconfig
+      files for correct behavior.
+
+     SoC
```
indentation fixed, and list sorted alphabetically.

tejlmand replied:
> the rendered page looks very funny:

Even though I have forced pushed with alphabetically sorted list and indentation fixed, the doc build output has not been updated.

carlescufi replied:
This is fixed now as far as I can tell:
https://builds.zephyrproject.io/zephyr/pr/69802/docs/hardware/porting/board_porting.html#terminology

iuliana-prodan commented on doc/hardware/porting/board_porting.rst:
```diff
@@ -157,19 +157,19 @@ not always used.
+   | :ref:`mimx8mp_evk <imx8mp_evk>`            | mimx8m/a53            | i.MX8M Plus | i.MXM8M A53   | NXP i.MX      | Arm Cortex-A53 | Arm64        |
    |                                            +-----------------------+-------------+---------------+---------------+----------------+--------------+
-   |                                            | /mimx8m/m7            | i.MX8M Plus | i.MXM8MM M4   | NXP i.MX      | Arm Cortex-M7  | Arm          |
+   |                                            | mimx8m/m7             | i.MX8M Plus | i.MXM8MM M4   | NXP i.MX      | Arm Cortex-M7  | Arm          |
```
The SoC Series is wrong here - it should be i.MX8M M7 - https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/nxp/imx/soc.yml#L32

iuliana-prodan replied:
Also, I would add here another entry for other arch, like Xtensa:
HiFi4 | Xtensa |  mimx8m/adsp | i.MX8M Plus | i.MX8M ADSP | NXP i.MX | HIFi4

iuliana-prodan commented on doc/hardware/porting/board_porting.rst:
```diff
@@ -157,19 +157,19 @@ not always used.
+   |                                            | nrf5340/cpunet        | nRF5340     | nRF53         | Nordic nRF    | Arm Cortex-M33 | Arm          |
    +--------------------------------------------+-----------------------+-------------+---------------+---------------+----------------+--------------+
-   | :ref:`mimx8mp_evk <imx8mp_evk>`            | /mimx8m/a53           | i.MX8M Plus | i.MXM8M A53   | NXP i.MX      | Arm Cortex-A53 | Arm64        |
+   | :ref:`mimx8mp_evk <imx8mp_evk>`            | mimx8m/a53            | i.MX8M Plus | i.MXM8M A53   | NXP i.MX      | Arm Cortex-A53 | Arm64        |
```
s/i.MXM8M A53/i.MX8M A53

tejlmand replied:
see https://github.com/zephyrproject-rtos/zephyr/pull/69802#discussion_r1514519675

iuliana-prodan commented on doc/hardware/porting/board_porting.rst:
```diff
@@ -157,19 +157,19 @@ not always used.
+   |                                            | nrf5340/cpunet        | nRF5340     | nRF53         | Nordic nRF    | Arm Cortex-M33 | Arm          |
    +--------------------------------------------+-----------------------+-------------+---------------+---------------+----------------+--------------+
-   | :ref:`mimx8mp_evk <imx8mp_evk>`            | /mimx8m/a53           | i.MX8M Plus | i.MXM8M A53   | NXP i.MX      | Arm Cortex-A53 | Arm64        |
+   | :ref:`mimx8mp_evk <imx8mp_evk>`            | mimx8m/a53            | i.MX8M Plus | i.MX8M        | NXP i.MX      | Arm Cortex-A53 | Arm64        |
```
Is `mimx8m` the correct identifier?
In boards/ we have `mimx8ml8`.
See:
A53 -  https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/nxp/imx8mp_evk/imx8mp_evk_mimx8ml8_a53.yaml#L7
M7 - https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/nxp/imx8mp_evk/imx8mp_evk_mimx8ml8_m7.yaml#L7

carlescufi replied:
Done

nashif commented on doc/hardware/porting/board_porting.rst:
```diff
@@ -67,27 +67,27 @@ complete conversion reference.
 
-Board and board identifiers
-***************************
+Board names, identifiers and targets
```
qualifiers?

carlescufi replied:
Sorry, this is not ready for review yet. Working on it now

carlescufi committed:
doc: add board terminology section to board porting guide

Add a terminology section to the board porting guide.
This provides a single location where terminologies used for boards can
be found and linked to.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

ajarmouni-st commented on doc/glossary.rst:
```diff
@@ -85,12 +128,30 @@ Glossary of Terms
+   variant
+      In the context of :term:`board qualifiers`, a variant designates a
+      particular type or configuration of a build for a combination of :term:`SoC`
+      and :term:`CPU cluster`. Commons uses of the variant concept include
```
```suggestion
      and :term:`CPU cluster`. Common uses of the variant concept include
```

tejlmand commented on doc/hardware/porting/board_porting.rst:
```diff
@@ -65,112 +65,128 @@ complete conversion reference.
-Examples of such use-cases are:
+- :term:`board`, which has one or more
+- :term:`SoC`, each of which optionally belong to a
+- :term:`SoC series`, which in turn may optionally belong to a
+- :term:`SoC family`, which has one or more
+- :term:`CPU Cluster`, each containing one or more
```
minor nit, this could give the impression that it is the SoC family which defines / has one or more CPU cluster.

In fact it is the SoC which defines the CPU cluster, not the family.

carlescufi replied:
Well spotted.

tejlmand commented on doc/hardware/porting/board_porting.rst:
```diff
@@ -65,112 +65,128 @@ complete conversion reference.
 
-Zephyr's hardware support hierarchy has these layers, from most to least
-specific:
+   {board name}[@{revision}][/{SoC}][/{CPU cluster}][/{variant}]
```
to me this indicates that you can have a CPU cluster or variant without a SoC.

Should this have been:
```suggestion
   {board name}[@{revision}][/{SoC}[/{CPU cluster}][/{variant}]]
```

carlescufi replied:
Indeed

tejlmand reviewed (commented):
LGTM, a +1 from here.

Two minor observations where I think things could be misinterpreted for newcomers.

tejlmand reviewed (commented):
LGTM, a +1 from here.

carlescufi reviewed (dismissed):
+1 from @tejlmand

nordicjm commented on doc/glossary.rst:
```diff
@@ -85,12 +128,30 @@ Glossary of Terms
+      as well as peripherals and memory.
+
+   SoC family
+      One or more :term:`SoCs <SoC>` or :term:`SoC series` that share enough
```
Should there be a term for extended board?

carlescufi replied:
Not sure what you mean by that?

nordicjm replied:
This feature was added some time ago https://github.com/zephyrproject-rtos/zephyr/pull/57390/files might make sense to have the term added to the glossary

carlescufi replied:
ooh right. The glossary needs tons of updates, among them this one :)  I will improve the glossary soon with another PR

nordicjm commented on doc/hardware/porting/board_porting.rst:
```diff
@@ -65,112 +65,128 @@ complete conversion reference.
+   +--------------------------------------------+--------------------------+-------------+--------------------+--------------------+----------------+----------------------+
+   | :term:`board name`                         | :term:`board qualifiers` | :term:`SoC` | :term:`SoC Series` | :term:`SoC family` | CPU core       | :term:`architecture` |
+   +============================================+==========================+=============+====================+====================+================+======================+
+   | :ref:`nrf52dk <nrf52dk_nrf52832>`          | nrf52832                 | nRF52832    | nRF52              | Nordic nRF         | Arm Cortex-M4  | Arm Cortex-M         |
+   +--------------------------------------------+--------------------------+-------------+--------------------+--------------------+----------------+----------------------+
+   | :ref:`frdm_k64f <frdm_k64f>`               | mk64f12                  | MK64F12     | Kinetis K6x        | NXP Kinetis        | Arm Cortex-M4  | Arm Cortex-M         |
+   +--------------------------------------------+--------------------------+-------------+--------------------+--------------------+----------------+----------------------+
+   | :ref:`rv32m1_vega <rv32m1_vega>`           | openisa_rv32m1/ri5cy     | RV32M1      | (Not used)         | (Not used)         | RI5CY          | RISC-V               |
+   +--------------------------------------------+--------------------------+-------------+--------------------+--------------------+----------------+----------------------+
+   | :ref:`nrf5340dk <nrf5340dk_nrf5340>`       | nrf5340/cpuapp           | nRF5340     | nRF53              | Nordic nRF         | Arm Cortex-M33 | Arm Cortex-M         |
+   |                                            +--------------------------+-------------+--------------------+--------------------+----------------+----------------------+
+   |                                            | nrf5340/cpunet           | nRF5340     | nRF53              | Nordic nRF         | Arm Cortex-M33 | Arm Cortex-M         |
+   +--------------------------------------------+--------------------------+-------------+--------------------+--------------------+----------------+----------------------+
+   | :ref:`mimx8mp_evk <imx8mp_evk>`            | mimx8ml8/a53             | i.MX8M Plus | i.MX8M             | NXP i.MX           | Arm Cortex-A53 | Arm64                |
+   |                                            +--------------------------+-------------+--------------------+--------------------+----------------+----------------------+
+   |                                            | mimx8ml8/m7              | i.MX8M Plus | i.MX8M             | NXP i.MX           | Arm Cortex-M7  | Arm Cortex-M         |
```
architecture isn't really right, M4 would be armv7, M33 would be armv8

carlescufi replied:
yes, good point. I will fix that.

henrikbrixandersen commented on doc/hardware/porting/board_porting.rst:
```diff
@@ -65,112 +65,128 @@ complete conversion reference.
-architectures then those must be modelled as multiple clusters, where all CPU
-cores within a cluster is having the same architecture.
-It is possible to have only a single CPU core within a CPU cluster.
+Zephyr's hardware support hierarchy has the following levels, from most to least
+specific:
```
Not sure I agree that the list goes from "most to least specific", but it's not a blocker.

henrikbrixandersen commented on doc/hardware/porting/board_porting.rst:
```diff
@@ -65,112 +65,128 @@ complete conversion reference.
+   +--------------------------------------------+--------------------------+-------------+--------------------+--------------------+----------------+----------------------+
+   | :term:`board name`                         | :term:`board qualifiers` | :term:`SoC` | :term:`SoC Series` | :term:`SoC family` | CPU core       | :term:`architecture` |
+   +============================================+==========================+=============+====================+====================+================+======================+
+   | :ref:`nrf52dk <nrf52dk_nrf52832>`          | nrf52832                 | nRF52832    | nRF52              | Nordic nRF         | Arm Cortex-M4  | Arm Cortex-M         |
+   +--------------------------------------------+--------------------------+-------------+--------------------+--------------------+----------------+----------------------+
+   | :ref:`frdm_k64f <frdm_k64f>`               | mk64f12                  | MK64F12     | Kinetis K6x        | NXP Kinetis        | Arm Cortex-M4  | Arm Cortex-M         |
+   +--------------------------------------------+--------------------------+-------------+--------------------+--------------------+----------------+----------------------+
+   | :ref:`rv32m1_vega <rv32m1_vega>`           | openisa_rv32m1/ri5cy     | RV32M1      | (Not used)         | (Not used)         | RI5CY          | RISC-V               |
+   +--------------------------------------------+--------------------------+-------------+--------------------+--------------------+----------------+----------------------+
+   | :ref:`nrf5340dk <nrf5340dk_nrf5340>`       | nrf5340/cpuapp           | nRF5340     | nRF53              | Nordic nRF         | Arm Cortex-M33 | Arm Cortex-M         |
+   |                                            +--------------------------+-------------+--------------------+--------------------+----------------+----------------------+
+   |                                            | nrf5340/cpunet           | nRF5340     | nRF53              | Nordic nRF         | Arm Cortex-M33 | Arm Cortex-M         |
+   +--------------------------------------------+--------------------------+-------------+--------------------+--------------------+----------------+----------------------+
+   | :ref:`mimx8mp_evk <imx8mp_evk>`            | mimx8ml8/a53             | i.MX8M Plus | i.MX8M             | NXP i.MX           | Arm Cortex-A53 | Arm64                |
+   |                                            +--------------------------+-------------+--------------------+--------------------+----------------+----------------------+
+   |                                            | mimx8ml8/m7              | i.MX8M Plus | i.MX8M             | NXP i.MX           | Arm Cortex-M7  | Arm Cortex-M         |
```
"Arm Cortex-M" isn't really the architecture in Zephyr terminology, though - is it?

carlescufi replied:
Yep, see the comment above

kartben reviewed (changes requested):
quick NACK on the size of the PNG file, while I do a more thorough review. 
209 KB does't sound right at all for such a simple file. Options: running it through optipng, look at WEBP format, or even straight SVG (which would likely also preserve the drawio metadata to allow for editing should it be needed)

See below (note that setting a smaller width (here 800px, but 500 might be even more appropriate since that's your target in the actual documentation page) would be better done directly when exporting from the tool you used, as it would minimize bluriness.

```
cwebp board-terminology.png -resize 800 0 -o out.webp                                           INT ✘  11:20:44  
Saving file 'out.webp'
File:      board-terminology.png
Dimension: 800 x 239
Output:    13920 bytes Y-U-V-All-PSNR 41.66 43.50 45.59   42.40 dB
           (0.58 bpp)
...
```

kartben commented on doc/hardware/porting/board_porting.rst:
```diff
@@ -65,112 +65,128 @@ complete conversion reference.
-omitted, that is ``plank//first`` and ``plank//second`` is an identical short
-form.
+   {board name}[@{revision}][/{board qualifiers}]
 
-.. _hw_support_hierarchy:
+which can be extended to:
 
-Boards, SoCs, etc.
-******************
+.. code-block:: console
 
-Zephyr's hardware support hierarchy has these layers, from most to least
-specific:
+   {board name}[@{revision}][/{SoC}[/{CPU cluster}][/{variant}]]
```
The combination of `[ ]`s and `{ }`s makes this really hard to parse visually IMO
Please consider the following:

```
Formally this can also be seen as
:samp:`{board name}[@{revision}][/{board qualifiers}]`, which can be extended to
:samp:`{board name}[@{revision}][/{SoC}[/{CPU cluster}][/{variant}]]`.
```

<img width="854" alt="image" src="https://github.com/zephyrproject-rtos/zephyr/assets/128251/ff9ffcb5-36d2-40cc-8a73-e488b454b7cc">

kartben commented on doc/hardware/porting/board_porting.rst:
```diff
@@ -65,112 +65,128 @@ complete conversion reference.
-- CPU core: a particular CPU in an architecture
-- Architecture: an instruction set architecture
+If a board contains only one single-core SoC, then the SoC can be omitted from
+the board target. This implies that, if the board does not define any board
```
```suggestion
the board target. This implies that if the board does not define any board
```

kartben reviewed (commented):
looking good, thanks for taking care of the image size @carlescufi. Another additional comment for improved readability, please.
